core-api/src/lib.rs

```diff
@@ -107,24 +93,30 @@ pub trait Core: Send + Sync {
     /// [Core::poll_workflow_activation]. If the workflow had any existing outstanding activations,
     /// such activations are invalidated and subsequent completions of them will do nothing and log
     /// a warning.
-    fn request_workflow_eviction(&self, task_queue: &str, run_id: &str);
+    fn request_workflow_eviction(&self, run_id: &str);
 
-    /// Returns core's instance of the [ServerGatewayApis] implementor it is using.
+    /// Returns this worker's instance of the [ServerGatewayApis] implementor it is using.
     fn server_gateway(&self) -> Arc<dyn ServerGatewayApis + Send + Sync>;
 
     /// Initiates async shutdown procedure, eventually ceases all polling of the server and shuts
-    /// down all registered workers. [Core::poll_workflow_activation] should be called until it
+    /// down this worker. [Worker::poll_workflow_activation] should be called until it
     /// returns [PollWfError::ShutDown] to ensure that any workflows which are still undergoing
     /// replay have an opportunity to finish. This means that the lang sdk will need to call
-    /// [Core::complete_workflow_activation] for those workflows until they are done. At that point,
-    /// the lang SDK can end the process, or drop the [Core] instance, which will close the
+    /// [Worker::complete_workflow_activation] for those workflows until they are done. At that point,
+    /// the lang SDK can end the process, or drop the [Worker] instance, which will close the
     /// connection.
     async fn shutdown(&self);
 
-    /// Shut down a specific worker. Will cease all polling on the task queue and future attempts
-    /// to poll that queue will return [PollWfError::NoWorkerForQueue].
-    async fn shutdown_worker(&self, task_queue: &str);
+    /// Completes shutdown and frees all resources. You should avoid simply dropping workers, as
+    /// this does not allow async tasks to report any panics that may have occurred cleanly.
+    ///
+    /// This should be called only after [Worker::shutdown] has resolved.
```

Comment:
Does `shutdown` return before shutdown complete (i.e. before it receives `complete_workflow_activation`s for outstanding activations)? If not, is there a good use case to separate `shutdown` and `finalize_shutdown`? If `shutdown` does return before workflow shutdown is technically all done, ignore me and leave them separate.

Reply:
It waits until everything is genuinely complete. The use case for `finalize_shutdown` is it actually consumes the worker struct and is able to free the inner resources like pollers in an ordered way. EX: Checking that spawned tasks did not panic.

Basically this is because you cannot do `async` things in `impl Drop`